### ISSUE TITLE
Detect image type based on content instead of file extension 

### ISSUE THREAD

**Issue by JalinWang** [typst/typst#3092]
### Description

Inspired by https://github.com/typst/typst/issues/3079, I find that currently, Typst tries to determine the image type by file extension first and then falls back to content-based guessing.

https://github.com/typst/typst/blob/e215f22965760bcc5df3204243597825e99b3fe7/crates/typst/src/visualize/image/mod.rs#L165-L176

However, many other libraries detect the image type based on the content by default, e.g., Pillow in Python. [ref](https://pillow.readthedocs.io/en/stable/handbook/image-file-formats.html#image-file-formats)
> The open() function identifies files from their contents, not their names, but the save() method looks at the name to determine which format to use, unless the format is given explicitly.

This is also a code of UNIX-like systems, i.e., do not rely on the file extension. [ref](https://www.hostinger.com/tutorials/linux-file-command/) ; [file command](https://manned.org/file.1)

Should we do it the same way? 
BTW, I'm not sure if it's okay to post here or should be put into Discussion. Please point it out if I miss something.

### Use Case

1. When extension is not correct https://github.com/typst/typst/issues/3079;
2. Keep consistency with other libraries like Pillow in Python;
3. Do something similar to the `file` command in Unix-like systems

**Comment by PgBiel** [comment]
I think a more useful request would be to allow overriding whatever the built-in detection process does, e.g. `image(..., format: "png")`.

**Comment by PgBiel** [comment]
Alternatively, one could add content-based detection as a fallback if the extension-based detection leads to an error.

**Comment by LaurenzV** [comment]
By the way, there is the `imagesize` crate which seems to be relatively lightweight and has a [method to detect the type of an image](https://docs.rs/imagesize/latest/imagesize/fn.image_type.html). Might be a bit more reliable than our current approach of looking at "magic bytes", although I haven't tested it.

**Comment by laurmaedje** [comment]
> By the way, there is the imagesize crate which seems to be relatively lightweight and has a [method to detect the type of an image](https://docs.rs/imagesize/latest/imagesize/fn.image_type.html). Might be a bit more reliable than our current approach of looking at "magic bytes", although I haven't tested it.

Right now, we use the `image` crate. From a quick look `imagesize` also seems to look at the first few bytes, so what's the difference?

**Comment by LaurenzV** [comment]
True, nevermind, I looked at the `size` function instead of the `matches` function, which also seems to mainly look at the first few bytes. My bad.

**Comment by laurmaedje** [comment]
Currently, Typst does the auto-detection if it doesn't know the extension. Upon further thought that seems like a backwards incompatibility hazard. If I have a `.png` JPEG file, it would detect as PNG and fail to decode, but I have a `.webp` JPEG file, Typst wouldn't know the extension and decode as JPEG. This would then break if we add WEBP support in the future. Only doing content-based detection when no file extension is present would be better in my opinion.

Adding a hint when the extension specifies one image format, but detection determines it as another seems fine, though not useful very often. I think expanding it to any file type is overkill, and would probably also require a new dependency.

**Comment by jayvdb** [comment]
fwiw, https://github.com/image-rs/image/issues/2368 is looking to add sniffing to `image`
